cmd/zsysd/completion.go

```diff
@@ -11,18 +12,58 @@ import (
 func installCompletionCmd(rootCmd *cobra.Command) {
 	prog := rootCmd.Name()
 	var completionCmd = &cobra.Command{
-		Use:   "completion",
-		Short: i18n.G("Generates bash completion scripts"),
-		Long: fmt.Sprintf(i18n.G(`To load completion run
+		Use:   "completion [bash|zsh|powershell]",
+		Short: i18n.G("Generates completion scripts"),
+		Long: strings.ReplaceAll(i18n.G(`To load completions:
 
-. <(%s completion)
+Bash:
 
-To configure your bash shell to load completions for each session add to your ~/.bashrc or ~/.profile:
+  $ source <(#prog# completion bash)
 
-. <(%s completion)
-`), prog, prog),
+  # To load completions for each session, execute once:
+  # Linux:
+  $ #prog# completion bash > /etc/bash_completion.d/#prog#
+  # macOS:
+  $ #prog# completion bash > /usr/local/etc/bash_completion.d/#prog#
+
+Zsh:
+
+  # If shell completion is not already enabled in your environment,
+  # you will need to enable it.  You can execute the following once:
+
+  $ echo "autoload -U compinit; compinit" >> ~/.zshrc
+
+  # To load completions for each session, execute once:
+  $ #prog# completion zsh > "${fpath[1]}/_#prog#"
+
+  # You will need to start a new shell for this setup to take effect.
+
+PowerShell:
+
+  PS> #prog# completion powershell | Out-String | Invoke-Expression
+
+  # To load completions for every new session, run:
+  PS> #prog# completion powershell > #prog#.ps1
+  # and source this file from your PowerShell profile.
+`), "#prog#", prog),
+		DisableFlagsInUseLine: true,
+		ValidArgs:             []string{"bash", "zsh", "powershell"},
```

Comment:
Does this still have an effect with cobra.MaximumNArgs(1)? (any validation done by it)

Reply:
Well, it does not deal with the validation, but it does provide completion.
Now, that might not be important (as this is run when completion doesn't work) - so I can remove it.
Validation is handled in the code by the `default` clause of the `switch`.

Let me know what would be preferred here.

Reply:
If this provides completion, this is a good reason enough to keep it :)

Reply:
on a related topic, any idea how (from the Run code) I can display the '--help' usage output, I've looked at the cobra docs but can't seem to find the solution.

Reply:
I want the usage output to show when the detected shell is blank or unknown

Reply:
Yes! You can get the command from the closure, and call `.Usage()` on it (see https://pkg.go.dev/github.com/spf13/cobra#Command.Usage). I think you can just do it quite elegantly in your switch in the default() case.
Then, after handling the error, just call `os.Exit(1)`.

Is there anything else blocking or did you misread you?